File: BIB/cocos2dx/MarmeladeGAFPlayer/Library/Sources/GAFTextureAtlas.cpp


#include "cocoagaf/GAFTexture.h"
#include "cocoagaf/GAFImage.h"
#include "cocoagaf/GAFInteger.h"
#include "cocoagaf/GAFString.h"
#include "cocoagaf/GAFArray.h"
#include "cocoagaf/GAFDictionary.h"

#include "GAFTextureAtlas.h"
#include "GAFTextureAtlasElement.h"
#include "GAFData.h"
#include "GAFAsset.h"

#include "CCJSONData.h"

#include <algorithm>

static const char * kAtlasesKey     = "atlases";
static const char * kElementsKey     = "elements";
static const char * kSourcesKey      = "sources";
static const char * kSourceKey       = "source";
static const char * kCSFKey          = "csf";
static const char * kTextureAtlasKey = "textureAtlas";


namespace GAF
{

GAFTextureAtlas::GAFTextureAtlas()
:
_loaded(false),
_images(NULL),
_textures(NULL),
_elements(NULL)
{
}

GAFTextureAtlas::~GAFTextureAtlas()
{
	CC_SAFE_RELEASE(_images);
	CC_SAFE_RELEASE(_textures);
	CC_SAFE_RELEASE(_elements);
}

GAFTextureAtlas * GAFTextureAtlas::create(const char * aTexturesDirectory, CCDictionary * aTextureAtlasConfigDictionary)
{
	GAFTextureAtlas *pRet = new GAFTextureAtlas();
    if (pRet && pRet->init(aTexturesDirectory, aTextureAtlasConfigDictionary))
    {
		pRet->autorelease();
        return pRet;
    }
    CC_SAFE_DELETE(pRet);
    return NULL;
}

static bool compare_atlases(const void* p1, const void* p2)
{
	CCDictionary* a1 = (CCDictionary*)p1;
	CCDictionary* a2 = (CCDictionary*)p2;
	CCInteger * id1 = (CCInteger*)a1->objectForKey("id");
	CCInteger * id2 = (CCInteger*)a2->objectForKey("id");
	GAFAssert(id1 && id2, "id parameters must be valid");
	return id1->getValue() < id2->getValue();
}


bool GAFTextureAtlas::init(const char * aTexturesDirectory, CCDictionary * aTextureAtlasConfigDictionary)
{
	if (!aTexturesDirectory || !aTextureAtlasConfigDictionary)
	{
		GAFLOG("parameters should not equal to nil");
		return false;
	}
	CCArray * atlasesInfo = (CCArray *)aTextureAtlasConfigDictionary->objectForKey(kAtlasesKey);
	
	std::sort(atlasesInfo->vect().begin(),
			  atlasesInfo->vect().end(),
			  compare_atlases);
	
	CC_SAFE_RELEASE(_images);
	_images = new CCArray();
	
	for (unsigned int i = 0; i < atlasesInfo->count(); ++i)
	{
		CCDictionary * atlasInfo = (CCDictionary*) atlasesInfo->objectAtIndex(i);
		int desiredCsf = GAFAsset::desiredCsf();
		CCArray * sources = (CCArray*) atlasInfo->objectForKey(kSourcesKey);
		std::string source;
		for (unsigned int j = 0; j < sources->count(); ++j)
		{
			CCDictionary * csfdict = (CCDictionary *) sources->objectAtIndex(j);
			CCNumber * scsf = (CCNumber*) csfdict->objectForKey(kCSFKey);
			int csf = scsf->getIntValue();
			CCString * s = (CCString*) csfdict->objectForKey(kSourceKey);
			if (1 == csf)
			{
				source = s->getCString();
			}
			if (csf == desiredCsf)
			{
				source = s->getCString();
				break;
			}
		}
		std::string td(aTexturesDirectory);
		std::string::size_type p = td.rfind('/');
		if (p == std::string::npos)
		{
			p = td.rfind('\\');
		}
		std::string path = td.substr(0, p + 1) + source;		
		CCImage * image = new CCImage();		
		image->initWithImageFile(path.c_str());
		_images->addObject(image);
		image->release();
	}
	loadElementsFromAnimationConfigDictionary(aTextureAtlasConfigDictionary);
	return true;
}

bool GAFTextureAtlas::loadElementsFromAnimationConfigDictionary(CCDictionary * aConfigDictionary)
{
    CCArray   * nElements     = (CCArray*)   aConfigDictionary->objectForKey(kElementsKey);
	
	if (nElements)
	{
		int s = nElements->count();
		
		CC_SAFE_RELEASE(_elements);
		_elements = CCDictionary::create();
		CC_SAFE_RETAIN(_elements);
		
		for (int i = 0; i < s; ++i)
		{
			CCObject * nElement = nElements->objectAtIndex(i);
			CCDictionary * dict = dynamic_cast<CCDictionary*>(nElement);
			if (dict)
			{
				GAFTextureAtlasElement * element = GAFTextureAtlasElement::create(dict);
				if (element)
				{
					_elements->setObject(element, element->name);
				}
			}
			else
			{
				GAFLOGERROR("Error when parsing texture atlas element JSON. Atlas element not of CCDictionary type");
			}
		}
	}
	else
	{
		GAFLOGERROR("Error when parsing texture atlas element JSON.");
		return false;
	}
	_loaded = true;
	return true;
}


CCDictionary * GAFTextureAtlas::elements()
{
	return _elements;
}

CCImage     * GAFTextureAtlas::image()
{
	if (_images && _images->count() > 0)
	{
		return (CCImage*) _images->objectAtIndex(0);
	}
	return NULL;
}

CCArray     * GAFTextureAtlas::images()
{
	return _images;
}

CCTexture2D * GAFTextureAtlas::texture()
{
	if (_textures && _textures->count() > 0)
	{
		return (CCTexture2D*) _textures->objectAtIndex(0);
	}
	return NULL;
}

CCArray * GAFTextureAtlas::textures()
{
	if (!_textures)
	{
		_textures = CCArray::createWithCapacity(_images->count());
		for (unsigned int i = 0; i < _images->count(); ++i)
		{
			CCTexture2D * texture = new CCTexture2D();
			CCImage * image = (CCImage*)_images->objectAtIndex(i);
			texture->initWithImage(image);
			_textures->addObject(texture);

			texture->release();
		}
		_textures->retain();
	}
	return _textures;
}

}

File: Flustered Pterodactyl/cocos2dx/Cocos2dGAFPlayer/Library/Sources/GAFTextureAtlas.m
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
//  GAFTextureAtlas.m
//  GAF Animation Library
//
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#import "GAFTextureAtlas.h"
#import "GAFTextureAtlasElement.h"
#import "CCTexture2D.h"
#import "Support/CCFileUtils.h"

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

static NSString * const kAtlasesKey     = @"atlases";
static NSString * const kElementsKey     = @"elements";
static NSString * const kSourcesKey      = @"sources";
static NSString * const kSourceKey       = @"source";
static NSString * const kCSFKey          = @"csf";
static NSString * const kTextureAtlasKey = @"textureAtlas";

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

@interface GAFTextureAtlas ()

@property (nonatomic, retain) NSMutableArray *images;
@property (nonatomic, retain) NSMutableArray *textures;

- (void)loadElementsFromAnimationConfigDictionary:(NSDictionary *)aConfigDictionary;

- (void)generateTexturesFromImages;

@end

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

@implementation GAFTextureAtlas

#pragma mark -
#pragma mark Properties

#pragma mark -
#pragma mark Initialization & Release

- (id)init
{
    self = [super init];
    if (nil != self)
    {
        self.images = [[[NSMutableArray alloc] initWithCapacity:1] ah_autorelease];
        self.textures = [[[NSMutableArray alloc] initWithCapacity:1] ah_autorelease];
        
        self.loaded = NO;
    }
    return self;
}

- (id)initWithTextureAtlasesDictionary:(NSDictionary *)aTextureAtlasesDictionary
          textureAtlasConfigDictionary:(NSDictionary *)aTextureAtlasConfigDictionary
                     keepImagesInAtlas:(BOOL)aKeepImagesInAtlas
{
    NSAssert(aTextureAtlasesDictionary != nil && aTextureAtlasConfigDictionary != nil, @"parameters should not equal to nil");
    if (aTextureAtlasesDictionary == nil)
    {
        CCLOGWARN(@"ERROR: initializing TextureAtlas. aTextureAtlasesDictionary not present");
        [self ah_release];
        return nil;
    }
    if (aTextureAtlasConfigDictionary == nil)
    {
        CCLOGWARN(@"ERROR: initializing TextureAtlas. aTextureAtlasConfigDictionary not present");
        [self ah_release];
        return nil;
    }
    
    if ((self = [self init]))
    {
        NSArray *atlasesInfo = aTextureAtlasConfigDictionary[kAtlasesKey];
        // Order by atlas id
        atlasesInfo = [atlasesInfo sortedArrayUsingComparator:
                       ^NSComparisonResult(id obj1, id obj2)
                       {
                           NSDictionary *info1 = (NSDictionary *)obj1;
                           NSDictionary *info2 = (NSDictionary *)obj2;
                           NSInteger id1 = [info1[@"id"] integerValue];
                           NSInteger id2 = [info2[@"id"] integerValue];
                           if (id2 > id1)
                               return NSOrderedAscending;
                           else if (id2 < id1)
                               return NSOrderedDescending;
                           else
                               return NSOrderedSame;
                       }];
        
        // Load textures from Data dictionary for each mentioned atlas
        if (atlasesInfo != nil && atlasesInfo.count > 0)
        {
            for (NSDictionary *atlasInfo in atlasesInfo)
            {
                NSArray *sources = atlasInfo[kSourcesKey];
                if (sources == nil)
                {
                    CCLOGWARN(@"ERROR: initializing sources. sources not present");
                    [self ah_release];
                    return nil;
                }
                
                int desiredCsf = 1;
                
                if ([[UIScreen mainScreen] respondsToSelector:@selector(displayLinkWithTarget:selector:)] &&
                    ([UIScreen mainScreen].scale == 2.0))
                {
                    desiredCsf = 2;
                }
                
                NSString * source = nil;
                int csf = 0;
                for (NSDictionary * csfdict in sources)
                {
                    csf = [csfdict[kCSFKey] intValue];
                    NSString * s = (NSString *)csfdict[kSourceKey];
                    if (1 == csf)
                    {
                        source = s;
                    }
                    if (csf == desiredCsf)
                    {
                        source = s;
                        break;
                    }
                }
                
                if (csf == 0)
                {
                    CCLOGWARN(@"ERROR: initializing sources.  csf not present");
                    [self ah_release];
                    return nil;
                }
                if (source == nil)
                {
                    CCLOGWARN(@"ERROR: initializing sources.  source not present");
                    [self ah_release];
                    return nil;
                }
                
                NSData *atlasTextureData = aTextureAtlasesDictionary[source];
                if (atlasTextureData != nil)
                {
                    UIImage *image = [[UIImage alloc] initWithData:atlasTextureData scale:csf];
                    if (image == nil)
                    {
                        CCLOGWARN(@"Cannot create UIImage for texture name(key) - %@", source);
                        [self ah_release];
                        return nil;
                    }
                    [self.images addObject:image];
                    [image ah_release];
                }
                else
                {
                    CCLOGWARN(@"Cannot find texture for name(key) - %@", source);
                    [self ah_release];
                    return nil;
                }
            }
        }
        
        [self loadElementsFromAnimationConfigDictionary:aTextureAtlasConfigDictionary];
        [self generateTexturesFromImages];
        if (!aKeepImagesInAtlas)
        {
            [self releaseImages];
        }
    }
    return self;
}

- (id)initWithTexturesDirectory:(NSString *)aTexturesDirectory
   textureAtlasConfigDictionary:(NSDictionary *)aTextureAtlasConfigDictionary
              keepImagesInAtlas:(BOOL)aKeepImagesInAtlas
{
    NSAssert(aTexturesDirectory != nil && aTextureAtlasConfigDictionary != nil, @"parameters should not equal to nil");
    if (aTexturesDirectory == nil)
    {
        CCLOGWARN(@"ERROR: initializing TextureAtlas. aTexturesDirectory not present");
        [self ah_release];
        return nil;
    }
    if (aTextureAtlasConfigDictionary == nil)
    {
        CCLOGWARN(@"ERROR: initializing TextureAtlas. aTextureAtlasConfigDictionary not present");
        [self ah_release];
        return nil;
    }
    
    if ((self = [self init]))
    {
        NSArray *atlasesInfo = aTextureAtlasConfigDictionary[kAtlasesKey];
        // Order by atlas id
        atlasesInfo = [atlasesInfo sortedArrayUsingComparator:
                       ^NSComparisonResult(id obj1, id obj2)
                       {
                           NSDictionary *info1 = (NSDictionary *)obj1;
                           NSDictionary *info2 = (NSDictionary *)obj2;
                           NSInteger id1 = [info1[@"id"] integerValue];
                           NSInteger id2 = [info2[@"id"] integerValue];
                           if (id2 > id1)
                               return NSOrderedAscending;
                           else if (id2 < id1)
                               return NSOrderedDescending;
                           else
                               return NSOrderedSame;
                       }];
        
        // Load textures for each mentioned atlas
        if (atlasesInfo != nil && atlasesInfo.count > 0)
        {
            for (NSDictionary *atlasInfo in atlasesInfo)
            {
                NSArray *sources = atlasInfo[kSourcesKey];
                if(sources == nil)
                {
                    CCLOGWARN(@"ERROR: initializing sources. 'sources' not present");
                    [self ah_release];
                    return nil;
                }
                int desiredCsf = 1;
                int realCsf = 1;
                
                if ([[UIScreen mainScreen] respondsToSelector:@selector(displayLinkWithTarget:selector:)] &&
                    ([UIScreen mainScreen].scale == 2.0))
                {
                    desiredCsf = 2;
                }
                
                int csf = 0;
                NSString * source = nil;
                for (NSDictionary * csfdict in sources)
                {
                    csf = [csfdict[kCSFKey] intValue];
                    NSString * s = (NSString *)csfdict[kSourceKey];
                    if (1 == csf)
                    {
                        source = s;
                    }
                    if (csf == desiredCsf)
                    {
                        source = s;
                        realCsf = csf;
                        break;
                    }
                }
                
                if (csf == 0)
                {
                    CCLOGWARN(@"ERROR: initializing sources. 'csf' not present");
                    [self ah_release];
                    return nil;
                }
                if (source == nil)
                {
                    CCLOGWARN(@"ERROR: initializing sources. 'source' not present");
                    [self ah_release];
                    return nil;
                }
                
                NSData *imageData = [NSData dataWithContentsOfFile:[aTexturesDirectory stringByAppendingPathComponent:source]];
                if (imageData == nil)
                {
                    CCLOGWARN(@"Cannot load imageData for name(key) - %@", source);
                    [self ah_release];
                    return nil;
                }
                UIImage *image = [[UIImage alloc] initWithData:imageData scale:realCsf];
                if (image == nil)
                {
                    CCLOGWARN(@"Cannot create UIImage for texture for name(key) - %@", source);
                    [self ah_release];
                    return nil;
                }
                [self.images addObject:image];
                [image ah_release];
            }
        }
        
        [self loadElementsFromAnimationConfigDictionary:aTextureAtlasConfigDictionary];
        [self generateTexturesFromImages];
        if (!aKeepImagesInAtlas)
        {
            [self releaseImages];
        }
    }
    return self;
}

- (id)initWithImages:(NSArray *)anImagesArray
       atlasElements:(NSDictionary *)anAtlasElements
    generateTextures:(BOOL)aGenerateTextures
{
    if (anImagesArray == nil)
    {
        CCLOGWARN(@"ERROR: initializing TextureAtlas. anImagesArray not present");
        [self ah_release];
        return nil;
    }
    if (anAtlasElements == nil)
    {
        CCLOGWARN(@"ERROR: initializing TextureAtlas. anAtlasElements not present");
        [self ah_release];
        return nil;
    }
    
    if ((self = [self init]))
    {
        self.images = [[anImagesArray mutableCopy] ah_autorelease];
        self.elements = [[anAtlasElements copy] ah_autorelease];
        
        self.loaded = YES;
        
        if (aGenerateTextures)
        {
            [self generateTexturesFromImages];
        }
    }
    return self;
}

- (void)dealloc
{
    [_images ah_release];
    [_textures ah_release];
    [_elements ah_release];

#if !__has_feature(objc_arc)
    [super dealloc];
#endif
}

#pragma mark -
#pragma mark Public methods

- (void)releaseImages
{
    if (self.images.count > 0)
    {
        if (self.images.count == self.textures.count)
        {
            [self.images removeAllObjects];
        }
        else
        {
            CCLOGWARN(@"GAFTextureAtlas images cannot be released as textures were not created, releasing of images will lead to complete image data loss");
        }
    }
}

#pragma mark -
#pragma mark Private methods

- (void)generateTexturesFromImages
{
    if (self.images.count == 0)
        return;
    
    if ([EAGLContext currentContext] == nil)
    {
        CCLOGWARN(@"Cannot create CCTexture2D in GAFTextureAtlas because there is no EAGLContext in thread %@", [NSThread currentThread]);
        return;
    }
    
    for (UIImage *image in self.images)
    {
        // kCCResolutioniPad kCCResolutioniPadRetinaDisplay // should not be used - it should work universally across the platforms
        CCTexture2D *texture = [[CCTexture2D alloc] initWithCGImage:image.CGImage resolutionType:kCCResolutioniPadRetinaDisplay];
        [self.textures addObject:texture];
        [texture ah_release];
    }
}

- (void)loadElementsFromAnimationConfigDictionary:(NSDictionary *)aConfigDictionary
{
    if (aConfigDictionary == nil)
    {
        CCLOGWARN(@"Error in loadElementsFromAnimationConfigDictionary. aConfigDictionary is nil.");
        return;
    }
    
    NSArray *nElements = aConfigDictionary[kElementsKey];
    
    if (nElements != nil)
    {
        @autoreleasepool
        {
            NSMutableDictionary *parsedElements = [[NSMutableDictionary alloc] initWithCapacity:[nElements count]];
            for (id nElement in nElements)
            {
                if ([nElement isKindOfClass:[NSDictionary class]])
                {
                    GAFTextureAtlasElement *element =
                        [[GAFTextureAtlasElement alloc] initWithDictionary:(NSDictionary *)nElement];
                    if (element != nil)
                    {
                        parsedElements[element.name] = element;
                        [element ah_release];
                    }
                }
                else
                {
                   CCLOGWARN(@"Error when parsing texture atlas element JSON. Atlas element not of NSDictionary type.");
                }
            }
            self.elements = parsedElements;
            [parsedElements ah_release];
        }
    }
    else
    {
        CCLOGWARN(@"Error when parsing texture atlas element JSON.");
        return;
    }
    
    self.loaded = YES;
}

@end